project-chip/connectedhomeip
# TV Android app - Add commissioning User Prompter to TV android app

#### Problem
- TV Android App does not have a UI support for [UserPrompter logic](https://github.com/project-chip/connectedhomeip/blob/master/examples/tv-app/android/java/AppImpl.cpp#L50-L100)
- End users are not able to confirm or decline commissioning request
- End users are not able to enter PIN to confirm commissioning request

#### Change overview
- Added JNI layer for Native C++ UserPrompter logic so that code can prompt end users with UI dialog 
- Added JNI layer to Android UserPrompter logic so that code can pass back user response (accept/decline) or pin code

#### Testing
Manually tested the flow using the instrucitons @chrisdecenzo shared with me:
1) Overwrite your src/platform/android/java/chip/platform/NsdManagerServiceResolver.java with the file provided
2) For full shell command functionality from casting app, replace your OperationalDeviceProxy.h/cpp with the file
3) Create a UDP port mapping from 5750 to 5650
telnet localhost 5554
auth XXXXX
redir add udp:5750:5650
redir add udp:5740:5640
4) run linux tv-casting-app. It won't discover the TV so you need to manually force a user-directed-commissioning request to the TV using the remapped port:
cast sendudc 127.0.0.1 5750
5) The TV should receive the UDC command and automatically commission the casting app. Watch tv-app logs for messages like:
------PROMPT USER: is requesting permission to cast to this TV, approve? [0x0000_0000,0x0000_0000,AB4E0808D7EFD3B3]
------PROMPT USER: please enter pin displayed in casting app
------PROMPT USER: commissioning success
Right now, tv-app is hardcoded to automatically commission since there is no current dialog box implemented.
